src/internal/m365/export.go

```diff
@@ -1,89 +1,30 @@
 package m365
 
 import (
-	"context"
-
-	"github.com/alcionai/clues"
-
-	"github.com/alcionai/corso/src/internal/data"
-	"github.com/alcionai/corso/src/internal/diagnostics"
-	"github.com/alcionai/corso/src/internal/m365/graph"
 	"github.com/alcionai/corso/src/internal/m365/service/groups"
 	"github.com/alcionai/corso/src/internal/m365/service/onedrive"
 	"github.com/alcionai/corso/src/internal/m365/service/sharepoint"
-	"github.com/alcionai/corso/src/internal/m365/support"
-	"github.com/alcionai/corso/src/pkg/backup/details"
+	"github.com/alcionai/corso/src/internal/operations/inject"
 	"github.com/alcionai/corso/src/pkg/control"
-	"github.com/alcionai/corso/src/pkg/export"
-	"github.com/alcionai/corso/src/pkg/fault"
-	"github.com/alcionai/corso/src/pkg/selectors"
+	"github.com/alcionai/corso/src/pkg/path"
 )
 
-// ProduceExportCollections exports data from the specified collections
-func (ctrl *Controller) ProduceExportCollections(
-	ctx context.Context,
-	backupVersion int,
-	sels selectors.Selector,
-	exportCfg control.ExportConfig,
+// NewServiceHandler returns an instance of a struct capable of running various
+// operations for a given service.
+func (ctrl *Controller) NewServiceHandler(
 	opts control.Options,
-	dcs []data.RestoreCollection,
-	stats *data.ExportStats,
-	errs *fault.Bus,
-) ([]export.Collectioner, error) {
-	ctx, end := diagnostics.Span(ctx, "m365:export")
-	defer end()
-
-	ctx = graph.BindRateLimiterConfig(ctx, graph.LimiterCfg{Service: sels.PathService()})
-	ctx = clues.Add(ctx, "export_config", exportCfg)
+	service path.ServiceType,
+) inject.ServiceHandler {
+	switch service {
+	case path.OneDriveService:
+		return onedrive.NewOneDriveHandler(opts)
 
-	var (
-		expCollections []export.Collectioner
-		status         *support.ControllerOperationStatus
-		deets          = &details.Builder{}
-		err            error
-	)
+	case path.SharePointService:
+		return sharepoint.NewSharePointHandler(opts)
 
-	switch sels.Service {
-	case selectors.ServiceOneDrive:
-		expCollections, err = onedrive.ProduceExportCollections(
-			ctx,
-			backupVersion,
-			exportCfg,
-			opts,
-			dcs,
-			deets,
-			stats,
-			errs)
-	case selectors.ServiceSharePoint:
-		expCollections, err = sharepoint.ProduceExportCollections(
-			ctx,
-			backupVersion,
-			exportCfg,
-			opts,
-			dcs,
-			ctrl.backupDriveIDNames,
-			deets,
-			stats,
-			errs)
-	case selectors.ServiceGroups:
-		expCollections, err = groups.ProduceExportCollections(
-			ctx,
-			backupVersion,
-			exportCfg,
-			opts,
-			dcs,
-			ctrl.backupDriveIDNames,
-			ctrl.backupSiteIDWebURL,
-			deets,
-			stats,
-			errs)
-
-	default:
-		err = clues.Wrap(clues.New(sels.Service.String()), "service not supported")
+	case path.GroupsService:
+		return groups.NewGroupsHandler(opts)
 	}
 
-	ctrl.incrementAwaitingMessages()
-	ctrl.UpdateStatus(status)
-
-	return expCollections, err
+	return &inject.BaseServiceHandler{}
```

Comment:
Not particularly a fan of deferring this to a base handler here.  The expectation, I assume, is for the base handler to error if used.  If we're in this code, we already know the caller intends to use the handler but has provided incorrect parameters.  An error up front feels more appropriate.